src/store/builder/index.ts

```diff
@@ -113,27 +117,113 @@ export const mutations: MutationTree<IBuilderState> = {
+    Object.assign(state.draggableForBlocksByUuid[uuid], { left: position.x, top: position.y })
+  },
+
+  updateToolBarHeight(state, height) {
+    state.toolbarHeight = height
   }
 }
 
 export const actions: ActionTree<IBuilderState, IRootState> = {
+  setBlockPositionTo({ state, commit, getters, dispatch, rootState, rootGetters }, { position: { x, y }, block }) {
+    // todo: ensure our vendor_metadata.io_viamo is always instantiated with builder // uiData props
+    // if (!block.vendor_metadata.io_viamo.uiData) {
+    //   defaultsDeep(block, {vendor_metadata: {io_viamo: {uiData: {xPosition: 0, yPosition: 0}}}})
+    //   Vue.observable(block.vendor_metadata.io_viamo.uiData)
+    // }
+
+    if (!includes(rootState.flow.selectedBlockUuids, block.uuid)) {
+      block.vendor_metadata.io_viamo.uiData.xPosition = x
+      block.vendor_metadata.io_viamo.uiData.yPosition = y
+    } else { // the block is selected
+      // Store UI Position
+      const initialPosition = {
+        x: clone(block.vendor_metadata.io_viamo.uiData.xPosition),
+        y: clone(block.vendor_metadata.io_viamo.uiData.yPosition),
+      } as IPosition
+
+      // Prepare translation
+      const translationDelta: IPosition = {
+        x: x - initialPosition.x,
+        y: y - initialPosition.y
+      }
+
+      let shouldReversePosition = false;
+      if (translationDelta.x > 0 /* moving to the right */ ||
+        rootGetters['flow/selectedBlockAtTheFurthestLeftPosition'].vendor_metadata.io_viamo.uiData.xPosition > 0 /* left space still available */) {
+        block.vendor_metadata.io_viamo.uiData.xPosition = x
+      } else {
+        translationDelta.x = 0
+        block.vendor_metadata.io_viamo.uiData.xPosition = initialPosition.x
+        shouldReversePosition = true
+      }
+
+      if(translationDelta.y > 0 /* moving to the top */ ||
+        rootGetters['flow/selectedBlockAtTheTopPosition'].vendor_metadata.io_viamo.uiData.yPosition - state.toolbarHeight > 0 /* top space still available */) {
+        block.vendor_metadata.io_viamo.uiData.yPosition = y
+      } else {
+        translationDelta.y = 0
+        block.vendor_metadata.io_viamo.uiData.yPosition = initialPosition.y
+        shouldReversePosition = true
+      }
+
+      // Reverse the draggable position
+      if (shouldReversePosition) {
+        commit('updateBlockDraggablePosition', {
+          uuid: block.uuid,
+          position: {
+            x: block.vendor_metadata.io_viamo.uiData.xPosition,
+            y: block.vendor_metadata.io_viamo.uiData.yPosition,
+          }
+        })
+      }
+
+      // Translate other selected blocks
+      const otherSelectedBlocks = filter(rootGetters['flow/selectedBlocks'], currentBlock => currentBlock.uuid !== block.uuid);
+      forEach(otherSelectedBlocks, async (currentBlock: IBlock) => {
+        let newPosition = {} as IPosition
+        await dispatch('setBlockPositionFromDelta', { delta: translationDelta, block: currentBlock }).then(
```

Comment:
Hi @safydy-r , a couple things on this one: I am feeling confused by the need to manage block and draggable separately. Can we wrap these two mutations into a single action (eg. `setBlockAndSyncDraggablePositionTo({pos})` or `setBlockAndSyncDraggablePositionFromDelta({delta})`) then just call that everywhere? I think we want them to be the same in most circumstances, right?